### PR TITLE
Add AssetPath::resolve_path and resolve_embed_path methods

### LINKED ISSUE #22239: Accept `&AssetPath` instead of `&str` in `AssetPath::resolve`
## What problem does this solve or what need does it fill?

`AssetPath::resolve` and `AssetPath::resolve_embed` accept `&str` and parse it internaly. But if you already have `AssetPath`, it's wasteful to call `.to_string` and let `resolve` parse it again:

https://github.com/bevyengine/bevy/blob/8813c18c8e9f0357870ae1f3190d324664598d9b/crates/bevy_asset/src/path.rs#L424-L428

## What solution would you like?

Accept `&AssetPath` instead of `&str` in `AssetPath::resolve` looks like a better API to me.

## What alternative(s) have you considered?

Add `_str` suffix to the current methods and and new methods that accept `&AcceptPath`.

### PR DESCRIPTION
**Pull request by Eyad3skr**
# Objective

Fixes #22239

## Solution

- Added `resolve_path(&AssetPath)` - equivalent to `resolve(&str)`
- Added `resolve_embed_path(&AssetPath)` - equivalent to `resolve_embed(&str)`
- Refactored `resolve_internal` to use shared `resolve_from_parts` helper
- Added unit tests for equivalence
- All semantic edge cases preserved (eg., label-only, or leading /)
- No behavior changes to existing methods

## Testing

- Added 2 new equivalence tests proving `resolve_path() ≡ resolve(&path.to_string())`
- All 20 tests pass: `cargo test -p bevy_asset --lib path::tests`
- Test covers: relative paths, labels, leading /, explicit sources, RFC 1808 embed behavior

### How to test

```
cargo test -p bevy_asset --lib path::tests
```

---

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Very clean work; thank you :)

**Comment by viridia** [comment]
My only concern is that this propagates a pattern that we've been wanting to get rid of for a long time (but have never got around to fixing).

Specifically, `AssetPath::resolve` should not use `PathBuf`. `PathBuf` does a lot of crazy things - like parsing windows drive letters - which simply don't apply to Bevy asset paths, and which are a potential source of bugs. It would be much better to have our own path resolution algorithm that operates on string slices, this would be fairly easy to write, easy to test, and the behavior would be predictable.

`PathBuf` makes sense when we are talking about actual file paths: that is, when an asset path gets transformed into a path used to actually open files. This happens inside the filesystem asset source, and is not part of the higher-level asset framework.

But asset paths themselves are not file paths, they are more like URLs.

**Comment by Eyad3skr** [comment]
Thanks @alice-i-cecile for the compliment and thanks @viridia for the detailed feedback, that actually makes sense.

I agree `AssetPath` is closer to a URL-like path than a filesystem path, and `PathBuf`/`std::path` normalization brings platform-specific behavior (e.g. Windows prefixes / drive letters) that shouldn’t apply at the asset-path layer.

This PR intentionally kept behavior identical to the existing `resolve` / `resolve_embed` implementation, and the new `resolve_*_path` methods reuse the same internal logic to avoid any semantic changes. That said, I understand the concern that refactoring here may further entrench the `PathBuf`-based approach.

I thought of 2 things could be done rn :

1. keep this PR scoped to the API improvement (accepting `&AssetPath` + equivalence tests), and I open a follow-up issue/PR to replace the resolution algorithm with a string-slice implementation, or
2. expand this PR to replace the `PathBuf` usage as part of the same change?

I’m happy to do either..

**Comment by viridia** [comment]
If we commit to doing a follow-up PR then it would alleviate my concerns.

File: crates/bevy_asset/src/path.rs [hunk 1]
```diff
@@ -416,6 +416,121 @@ impl<'a> AssetPath<'a> {
...
+    fn resolve_from_parts(
+        &self,
+        replace: bool,
+        source: Option<&str>,
+        rpath: &Path,
+        rlabel: Option<&str>,
+    ) -> AssetPath<'static> {
+        let mut base_path = PathBuf::from(self.path());
+        if replace && !self.path.to_str().unwrap().ends_with('/') {
+            // No error if base is empty
```
Comment (Shatur) [inline crates/bevy_asset/src/path.rs]: Why `(per RFC 1808)` in the comment was removed?

File: crates/bevy_asset/src/path.rs [hunk 2]
```diff
@@ -416,6 +416,121 @@ impl<'a> AssetPath<'a> {
         self.resolve_internal(path, true)
     }
 
+    /// Resolves `path` relative to `self`, using the same rules as [`AssetPath::resolve`],
+    /// but without reparsing from a string.
```
Comment (Shatur) [inline crates/bevy_asset/src/path.rs]: Maybe change the description of `resolve` instead and mention that it parses the asset path from the string?

**Review by Shatur — CHANGES_REQUESTED** [review]
I think we can sneak this into 0.18 since it's not a breaking change, but I'd probably rename old methods into `*_str` and keep path variants as "default" in a follow-up.

**Comment by Eyad3skr** [comment]
@viridia Tracking issue opened: #22420 . I’ll follow up with a PR replacing PathBuf resolution with a deterministic segment-based resolver.

**Review by viridia — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
@Eyad3skr if you can tackle @Shatur's comments briefly I'll be happy to merge this and wait for the follow-up issue to address the `PathBuf` concerns :)

**Comment by Eyad3skr** [comment]
@Shatur @alice-i-cecile On it

**Comment by Shatur** [comment]
I meant to rename in a follow-up PR some time later because this would potentially allow us to include this change in 0.18. But if it's easier for you to do it now - it's okay, not that important to have this in 0.18 :slightly_smiling_face:

**Comment by Eyad3skr** [comment]
@Shatur oh really, sorry didn't catch this up! anyway I'm up for the night dodging studying for data engineering final, it is ok.

and sorry for the many mentions, we are more than 2 in the discussion, just so I don't mix things up.

File: crates/bevy_asset/src/path.rs [hunk 3]
```diff
@@ -349,73 +349,197 @@ impl<'a> AssetPath<'a> {
         self.clone().into_owned()
     }
 
-    /// Resolves a relative asset path via concatenation. The result will be an `AssetPath` which
-    /// is resolved relative to this "base" path.
+    /// Resolves an already-parsed [`AssetPath`] relative to `self`.
```
Comment (Shatur) [inline crates/bevy_asset/src/path.rs]: I think it's implied that `AssetPath` stores a parsed representation.
```suggestion
    /// Resolves an [`AssetPath`] relative to `self`.
```

File: crates/bevy_asset/src/path.rs [hunk 4]
```diff
@@ -349,73 +349,197 @@ impl<'a> AssetPath<'a> {
...
+            self.resolve_from_parts(
+                false, // replace=false for resolve (not resolve_embed)
+                explicit_source,
+                path.path(),
+                path.label(),
+            )
+        }
+    }
+
+    /// Resolves an already‑parsed [`AssetPath`] relative to `self` using embedded (RFC 1808) semantics.
```
Comment (Shatur) [inline crates/bevy_asset/src/path.rs]:
```suggestion
    /// Resolves an [`AssetPath`] relative to `self` using embedded (RFC 1808) semantics.
```

File: crates/bevy_asset/src/path.rs [hunk 5]
```diff
@@ -349,73 +349,197 @@ impl<'a> AssetPath<'a> {
...
     ///
-    /// There are several cases:
+    /// This is the primary, allocation‑free API. Use the string‑based variant
+    /// [`AssetPath::resolve_embed_str`] when you need to parse from `&str`.
     ///
-    /// If the `path` argument begins with `#`, then it is considered an asset label, in which case
-    /// the result is the base path with the label portion replaced.
+    /// Semantics:
+    /// - Remove the “file portion” of the base before concatenation (unless the base ends with `/`).
```
Comment (Shatur) [inline crates/bevy_asset/src/path.rs]: Please, avoid using AI for writing descriptions :pray: 
```suggestion
    /// - Remove the "file portion" of the base before concatenation (unless the base ends with `/`).
```

File: crates/bevy_asset/src/path.rs [hunk 6]
```diff
@@ -349,73 +349,197 @@ impl<'a> AssetPath<'a> {
...
     ///
-    /// If the `path` argument begins with an asset source (ex: `http://`) then the entire base
-    /// path is replaced - the result is the source, path and label (if any) of the `path`
-    /// argument.
+    /// Use this when starting from a string. For an allocation‑free variant that
+    /// takes an already‑parsed path, see [`AssetPath::resolve`].
     ///
-    /// Otherwise, the `path` argument is considered a relative path. The result is concatenated
-    /// using the following algorithm:
+    /// Semantics:
```
Comment (Shatur) [inline crates/bevy_asset/src/path.rs]: I feel like we're duplicating the docs quite a bit. Maybe move things to non-string variants and just link to them?

**Review by Shatur — CHANGES_REQUESTED** [review]
It's ok, let's keep it this way.
But I think docs require a bit of polishing.

File: crates/bevy_asset/src/path.rs (hunk 5 above)
Comment (Eyad3skr) [reply crates/bevy_asset/src/path.rs]: you are right, thought that ai would make it better, I will go write it myself though it is my first time writing docs, sorry for that,

Comment (Shatur) [reply crates/bevy_asset/src/path.rs]: Just try to look at the docs and think what would you like to read to get the idea about the method.
AI generates very awkward descriptions.

File: crates/bevy_asset/src/path.rs [hunk 7]
```diff
@@ -349,73 +349,160 @@ impl<'a> AssetPath<'a> {
...
     }
 
-    /// Resolves a relative asset path via concatenation. The result will be an `AssetPath` which
-    /// is resolved relative to this "base" path.
+    /// Resolves an [`AssetPath`] relative to `self`.
+    ///
+    /// Use this when you already have an [`AssetPath`]. If you have a string, see
+    /// [`AssetPath::resolve_str`].
+    ///
+    /// Semantics (also for [`AssetPath::resolve_str`]):
```
Comment (Shatur) [inline crates/bevy_asset/src/path.rs]: I think it's better to mention it in `resolve_str` docs.
```suggestion
    /// Semantics:
```

File: crates/bevy_asset/src/path.rs [hunk 8]
```diff
@@ -349,73 +349,160 @@ impl<'a> AssetPath<'a> {
...
+    /// - If `path` is label-only (default source, empty path, label set), replace `self`’s label.
```
Comment (Shatur) [inline crates/bevy_asset/src/path.rs]:
```suggestion
    /// - If `path` is label-only (default source, empty path, label set), replace `self`'s label.
```

File: crates/bevy_asset/src/path.rs [hunk 9]
```diff
@@ -349,73 +349,160 @@ impl<'a> AssetPath<'a> {
         self.clone().into_owned()
     }
 
-    /// Resolves a relative asset path via concatenation. The result will be an `AssetPath` which
-    /// is resolved relative to this "base" path.
+    /// Resolves an [`AssetPath`] relative to `self`.
+    ///
+    /// Use this when you already have an [`AssetPath`]. If you have a string, see
+    /// [`AssetPath::resolve_str`].
```
Comment (Shatur) [inline crates/bevy_asset/src/path.rs]: How about just this in the end (after "semantics"):
```suggestion
    /// See also [`AssetPath::resolve_str`].
```

File: crates/bevy_asset/src/path.rs [hunk 10]
```diff
@@ -349,73 +349,160 @@ impl<'a> AssetPath<'a> {
...
+        let is_label_only = matches!(path.source(), AssetSourceId::Default)
+            && path.path().as_os_str().is_empty()
+            && path.label().is_some();
+
+        if is_label_only {
+            // Label-only: replace label on base path
+            self.clone_owned()
+                .with_label(path.label().unwrap().to_owned())
+        } else {
+            // Determine if the input has an explicit (non-default) source
```
Comment (Shatur) [inline crates/bevy_asset/src/path.rs]: I suggest to avoid duplicating the code in English. Use comments when the intent is not clear or when they add additional information. This one just duplicates what is already written.
Same for most of the comments.
```suggestion
```

File: crates/bevy_asset/src/path.rs [hunk 11]
```diff
@@ -349,73 +349,160 @@ impl<'a> AssetPath<'a> {
...
+                path.label(),
+            )
+        }
+    }
+
+    /// Parses `path` as an [`AssetPath`], then resolves it relative to `self`.
     ///
-    /// * The base path and the `path` argument are concatenated.
-    /// * Path elements consisting of "/." or "&lt;name&gt;/.." are removed.
+    /// For resolution semantics, see [`AssetPath::resolve`].
```
Comment (Shatur) [inline crates/bevy_asset/src/path.rs]: Here and for the other `*_str` variant.
```suggestion
    /// For more details see [`AssetPath::resolve`].
```

File: crates/bevy_asset/src/path.rs [hunk 12]
```diff
@@ -1015,11 +1072,141 @@ mod tests {
...
         let base = AssetPath::from("../../a/b.gltf");
         assert_eq!(
-            base.resolve("c.bin").unwrap(),
+            base.resolve_str("c.bin").unwrap(),
             AssetPath::from("../../a/b.gltf/c.bin")
         );
     }
 
+    #[test]
+    fn resolve_path_equivalence() {
```
Comment (Shatur) [inline crates/bevy_asset/src/path.rs]: I'd suggest to just extend each test below. I.e. in addition to `resolve_str`, just add the regular `resolve`. This scales much better.

**Review by Shatur — CHANGES_REQUESTED** [review]

File: crates/bevy_asset/src/path.rs [hunk 13]
```diff
@@ -349,73 +349,148 @@ impl<'a> AssetPath<'a> {
         self.clone().into_owned()
     }
 
-    /// Resolves a relative asset path via concatenation. The result will be an `AssetPath` which
-    /// is resolved relative to this "base" path.
+    /// Resolves an [`AssetPath`] relative to `self`.
+    ///
+    /// Use this when you already have an [`AssetPath`]. If you have a string, see
+    /// [`AssetPath::resolve_str`].
```
Comment (Shatur) [inline crates/bevy_asset/src/path.rs]: No longer needed (you link to `resolve_str` in the end).
```suggestion
```

File: release-content/migration-guides/assetpath-resolve-semantics.md [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+- `AssetPath::resolve(&str)` → `AssetPath::resolve_str(&str)`
+- `AssetPath::resolve_embed(&str)` → `AssetPath::resolve_embed_str(&str)`
+
+**New methods:**
+- `AssetPath::resolve(&AssetPath)` - allocation-free variant
+- `AssetPath::resolve_embed(&AssetPath)` - allocation-free variant
+
+## How do I migrate?
+
+**Before:**
```
Comment (viridia) [inline release-content/migration-guides/assetpath-resolve-semantics.md]: This seems backwards? I'm not sure I understand. `resolve_str` didn't exist in the "before" world.

File: release-content/migration-guides/assetpath-resolve-semantics.md [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+let base = AssetPath::parse("a/b.gltf");
+let rel = AssetPath::parse("c.bin");
+let resolved = base.resolve_str(&rel.to_string()).unwrap();
+```
+
+**After:**
+```rust
+let base = AssetPath::parse("a/b.gltf");
+let rel = AssetPath::parse("c.bin");
+let resolved = base.resolve(&rel); //Behavior
```
Comment (viridia) [inline release-content/migration-guides/assetpath-resolve-semantics.md]: Comment seems incomplete.

File: release-content/migration-guides/assetpath-resolve-semantics.md (hunk 14 above)
Comment (Eyad3skr) [reply release-content/migration-guides/assetpath-resolve-semantics.md]: idk, my pov is someone will migrate he/she will check which case applies for him and see (after) to know what should he/she do

File: release-content/migration-guides/assetpath-resolve-semantics.md (hunk 15 above)
Comment (Eyad3skr) [reply release-content/migration-guides/assetpath-resolve-semantics.md]: removed the comment at all due to linting error in CI in the prev commit

File: release-content/migration-guides/assetpath-resolve-semantics.md (hunk 14 above)
Comment (Eyad3skr) [reply release-content/migration-guides/assetpath-resolve-semantics.md]: maybe this part need more clarification, I will see about it

**Comment by abdlla3amer** [comment]
Wonderful analogy with a high-level end-to-end object oriented solution Mr. @Eyad3skr

**Review by viridia — APPROVED** [review]

**Comment by Eyad3skr** [comment]
anybody got an idea of why CI / markdownlint (pull_request) fails?

**Comment by alice-i-cecile** [comment]
Markdown errors can be found starting https://github.com/bevyengine/bevy/actions/runs/20803366790/job/59752575411?pr=22416#step:4:156; you need to expand the markdown section when looking at the job :)

Thanks for your work here; I'll merge this once CI is passing!

File: release-content/migration-guides/assetpath-resolve-semantics.md [hunk 16]
```diff
@@ -0,0 +1,54 @@
+---
+title: "AssetPath::resolve` and `resolve_embed` now take `&AssetPath"
+pull_requests: [22416]
+---
+
+# `AssetPath::resolve` and `resolve_embed` now take `&AssetPath`
```
Comment (Shatur) [inline release-content/migration-guides/assetpath-resolve-semantics.md]: I'd suggest to take a look at how other people write migration guides: https://github.com/bevyengine/bevy/tree/main/release-content/migration-guides
It's a bit verbose for such a simple change. I wouldn't even add "Before" and "After".

**Review by Shatur — CHANGES_REQUESTED** [review]

File: release-content/migration-guides/assetpath-resolve-semantics.md (hunk 16 above)
Comment (Shatur) [reply release-content/migration-guides/assetpath-resolve-semantics.md]: @Eyad3skr I didn't mean to remove the labels. I meant to remove the whole blocks with the before and after code.

Comment (Eyad3skr) [reply release-content/migration-guides/assetpath-resolve-semantics.md]: so migration part I just write it in plain English?

Comment (Shatur) [reply release-content/migration-guides/assetpath-resolve-semantics.md]: Yeah, just a few sentences should be enough to explain the breaking change. Just take a look at the link I shared above.

Comment (Eyad3skr) [reply release-content/migration-guides/assetpath-resolve-semantics.md]: passed it finally.. the linter was confused between the real title from template and my text because I used H1 after the title, so I reduced it to H2

Comment (Eyad3skr) [reply release-content/migration-guides/assetpath-resolve-semantics.md]: @Shatur Ok I will see now

Comment (Eyad3skr) [reply release-content/migration-guides/assetpath-resolve-semantics.md]: Done

Comment (Shatur) [reply release-content/migration-guides/assetpath-resolve-semantics.md]: I think it's still overly verbose. Take a look at the migration guides I linked to you. You don't need headings, just write in plain English what was done (no need to write "# What was done"), how to migrate and why. It's obvious that if you have a string, it's more convenient to use the `*_str` methods.

Comment (Eyad3skr) [reply release-content/migration-guides/assetpath-resolve-semantics.md]: fair enough, made it

File: release-content/migration-guides/assetpath-resolve-semantics.md [hunk 17]
```diff
@@ -0,0 +1,8 @@
+---
+title: "`AssetPath::resolve` and `resolve_embed` now take `&AssetPath`"
+pull_requests: [22416]
+---
+
+`AssetPath::resolve` and `AssetPath::resolve_embed` no longer accept `&str` and now take `&AssetPath` directly. The previous string-based APIs have been renamed to `resolve_str` and `resolve_embed_str`.
+
+This change avoids unnecessary string allocation and parsing when an `AssetPath` is already available. To migrate, pass an `AssetPath` directly to `resolve` or `resolve_embed`; when working with strings, use the corresponding `*_str` methods instead. No behavioral or semantic changes were introduced.
```
Comment (Shatur) [inline release-content/migration-guides/assetpath-resolve-semantics.md]: Almost perfect, last tiny change, can be applied directly from the browser :slightly_smiling_face: I think the last sentence doesn't add anything.
```suggestion
This change avoids unnecessary string allocation and parsing when an `AssetPath` is already available. To migrate, pass an `AssetPath` directly to `resolve` or `resolve_embed`; when working with strings, use the corresponding `*_str` methods instead.
```

Comment (Eyad3skr) [reply release-content/migration-guides/assetpath-resolve-semantics.md]: made it, the linter failed anyway, needed a trailing newline

Comment (Eyad3skr) [reply release-content/migration-guides/assetpath-resolve-semantics.md]: @Shatur btw as for the issue #22420, ig it needs to be labeled and see if my description is right so I can go on with it?

**Review by Shatur — APPROVED** [review]
Perfect now! Thank you.

File: release-content/migration-guides/assetpath-resolve-semantics.md (hunk 17 above)
Comment (Shatur) [reply release-content/migration-guides/assetpath-resolve-semantics.md]: I'll leave the guiding for https://github.com/bevyengine/bevy/issues/22420 to @viridia .

I want to note that `PathBuf` was already used and you just moved it into a separate method to re-use the logic for `AssetPath`. So you did a great job.
#22420 it's a separate, but related issue.

**Comment by Eyad3skr** [comment]
> Perfect now! Thank you.

Always! Thank you for the review, much appreciated!

File: release-content/migration-guides/assetpath-resolve-semantics.md (hunk 17 above)
Comment (Eyad3skr) [reply release-content/migration-guides/assetpath-resolve-semantics.md]: Perfect, waiting for any updates. I'm trying to catch up anyway so I got no problem getting assigned this one too, thanks for the review and your effort!

**Comment by Shatur** [comment]
@alice-i-cecile this PR includes a breaking change, so we probably have to remove it from 0.18 :cry:
I suggested this one for 0.18 originally, but the author also included a rename. I don't think it's worth backporting the change without the rename for 0.18.
